Adds a tooltip style menu component

kumar303 opened the pull request:
Fixes https://github.com/mozilla/addons-frontend/issues/3534

We need a proper tooltip component (especially to handle edge collisions) for the *flag review* UI. Here is just the tooltip part.

You can see it with this patch:

```diff
diff --git a/src/amo/components/AddonReviewListItem/index.js b/src/amo/components/AddonReviewListItem/index.js
index 71b97881..c43a5bff 100644
--- a/src/amo/components/AddonReviewListItem/index.js
+++ b/src/amo/components/AddonReviewListItem/index.js
@@ -18,8 +18,10 @@ import {
   showReplyToReviewForm,
 } from 'amo/actions/reviews';
 import Icon from 'ui/components/Icon';
+import ListItem from 'ui/components/ListItem';
 import LoadingText from 'ui/components/LoadingText';
 import Rating from 'ui/components/Rating';
+import TooltipMenu from 'ui/components/TooltipMenu';
 import DismissibleTextForm from 'ui/components/DismissibleTextForm';
 import type { UserReviewType } from 'amo/actions/reviews';
 import type { ReviewState } from 'amo/reducers/reviews';
@@ -274,6 +276,25 @@ export class AddonReviewListItemBase extends React.Component {
                 </a>
               ) : null
           }
+          <TooltipMenu
+            items={[
+              <ListItem key="flag-spam">
+                <a href="#flag-review">
+                  {i18n.gettext('This review is spam')}
+                </a>
+              </ListItem>,
+              <ListItem key="flag-inappropriate-language">
+                <a href="#flag-review">
+                  {i18n.gettext(
+                    'This review uses inappropriate language'
+                  )}
+                </a>
+              </ListItem>,
+            ]}
+            openerClass="AddonReviewListItem-control"
+            openerText={i18n.gettext('Flag')}
+            openerTitle={i18n.gettext('Flag this review')}
+          />
         </div>
         {errorHandler.renderErrorIfPresent()}
         {this.renderReply()}

```

Here is how it looks on desktop:

<img width="709" alt="screenshot 2017-10-18 12 06 32" src="https://user-images.githubusercontent.com/55398/31732336-d06fa69c-b3fd-11e7-8fd2-f2cc0bc7e236.png">
<img width="709" alt="screenshot 2017-10-18 12 12 26" src="https://user-images.githubusercontent.com/55398/31732344-d5396a28-b3fd-11e7-9ed0-c8cbdffd63f5.png">


Tooltips near the edge get adjusted so that they never render off screen:

<img width="324" alt="screenshot 2017-10-18 12 07 09" src="https://user-images.githubusercontent.com/55398/31732354-dd886c42-b3fd-11e7-9731-ec32c3f557a1.png">
<img width="324" alt="screenshot 2017-10-18 12 12 03" src="https://user-images.githubusercontent.com/55398/31732358-e2476b66-b3fd-11e7-908b-9b97846a5686.png">

muffinresearch commented:
The rc-tooltip lib looks to have an open issue re: keyboard accessibility. Can you access the links if you just use the keyboard? 

We'd probably need to go a different route if it's not possible to set a reason without using a mouse. E.g could the reason just be added to the form as a radio button instead?

kumar303 commented:
> could the reason just be added to the form as a radio button instead

I'm not sure. There will eventually be an *other* option which would close the menu and let the user enter some text.

willdurand commented:
> Can you access the links if you just use the keyboard?

I cannot access the links unfortunately. I can "tab" to reach the `Flag`, then `enter` to open the menu/tooltip (`space` does not work whereas it works for the dropdown menu). If I "tab" again, I go to the next `Flag`.

willdurand commented on src/ui/components/TooltipMenu/styles.scss:
```diff
@@ -0,0 +1,65 @@
+    a:active {
+      font-weight: normal;
+      text-decoration: none;
+    }
```
Why don't we have black links like the dropdown menu? It should be blue only when `:active`, `:focus`, or `:hover`.

kumar303 replied:
Good idea, thanks. I actually began this patch by simply extending the `DropdownMenu` but then I realized edge detection is something I didn't want to write from scratch.

willdurand commented:
<img width="447" alt="screen shot 2017-10-18 at 13 47 50" src="https://user-images.githubusercontent.com/217628/31717207-481af3a6-b40b-11e7-9482-c61bf8a31d34.png">
<img width="1392" alt="screen shot 2017-10-18 at 13 49 12" src="https://user-images.githubusercontent.com/217628/31717209-4841b586-b40b-11e7-8667-f8093de02342.png">

kumar303 commented on src/ui/components/TooltipMenu/index.js:
```diff
@@ -0,0 +1,80 @@
+  // TODO: put back this type definition when we upgrade Flow.
+  // It is triggering buggy React behavior.
+  // items: Array<ListItem>,
+  items: Array<React.Element<any>>,
```
It more or less gives the same error which I think is just a bug. Flow had a big overhaul of React support in a version newer than what we're on.

```
src/amo/components/AddonReviewListItem/index.js:285
285:               <ListItem key="flag-spam">
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^ React$Element. This type is incompatible with
 16:   items: Array<ListItem>,
                    ^^^^^^^^ ListItem. See: src/ui/components/TooltipMenu/index.js:16
```

kumar303 commented on src/ui/components/ListItem/index.js:
```diff
@@ -0,0 +1,30 @@
+  // By requiring a key, Flow would tell us if it was omitted accidentally.
+  // However, it is not possible to define `key` here since it's an
+  // internally hidden prop. See https://github.com/facebook/flow/issues/4202
+  // key: string | number,
```
Oh, whoops, I meant to delete it.

kumar303 commented on src/ui/components/ListItem/index.js:
```diff
@@ -0,0 +1,30 @@
+import makeClassName from 'classnames';
+import React from 'react';
+
+type PropsType = {|
```
You can always just use `any` for types that are hard to figure out. This would still provide better support than PropTypes.

willdurand replied:
ah no sorry, I was talking about the name of the Flow type sorry. No, we do not use the package `prop-types` anymore (and I do like Flow).

kumar303 replied:
Ha. My bad.

willdurand replied:
On IRC, we agreed on `Props` and `State`.

kumar303 commented on src/ui/components/TooltipMenu/index.js:
```diff
@@ -0,0 +1,80 @@
+    this.validateProps(props);
+  }
+
+  validateProps(props: PropsType) {
```
Yeah, good idea. I hope we can remove it soon. Maybe I'll find a moment to upgrade Flow.

kumar303 commented:
@willdurand could you take another look? I updated the patch (in the description) which had to change a bit. I fixed the RTL issue and somewhat address accessibility. The remaining issue is that you can't use the keyboard to select links but I think it will be better to follow up on in a new issue.

kumar303 commented:
Keyboard accessibility: https://github.com/mozilla/addons-frontend/issues/3556

willdurand commented on src/ui/components/TooltipMenu/styles.scss:
```diff
@@ -0,0 +1,76 @@
+      color: $blue-50;
+    }
+  }
+}
```
I would like to see this patch in this file, it does 2 things:

- it adds the `DropdownMenu` text style
- it adjusts the rendering on small screens

``` diff
diff --git a/src/ui/components/TooltipMenu/styles.scss b/src/ui/components/TooltipMenu/styles.scss
index 07a149a..aa00ed5 100644
--- a/src/ui/components/TooltipMenu/styles.scss
+++ b/src/ui/components/TooltipMenu/styles.scss
@@ -1,10 +1,16 @@
+@import "~amo/css/inc/vars";
 @import "~core/css/inc/mixins";
 @import "~ui/css/vars";
 
 .TooltipMenu {
   border: 0;
   border-radius: $border-radius-default;
-  width: 300px;
+  padding: 0 10px;
+  width: 100%;
+
+  @include respond-to(medium) {
+    width: 320px;
+  }
 }
 
 .TooltipMenu-opener {
@@ -60,10 +66,16 @@
       padding-top: 6px;
     }
 
+    a,
     a:link {
+      @include font-medium();
+
       color: $grey-90;
       cursor: pointer;
       font-size: $font-size-m-smaller;
+      line-height: 1.6;
+      margin: 0;
+      padding: 0;
       text-decoration: none;
     }
```

Example:

<img width="447" alt="screen shot 2017-10-18 at 21 06 35" src="https://user-images.githubusercontent.com/217628/31737564-40d45078-b448-11e7-8dee-2f33372f944f.png">

willdurand commented on src/ui/components/TooltipMenu/index.js:
```diff
@@ -0,0 +1,87 @@
+
+  validateProps(props: Props) {
+    const { _config } = this.props;
+    if (_config.get('isDevelopment') && props.items) {
```
Hopefully I can get rid of this after #3551 😈

willdurand commented on src/ui/components/ListItem/index.js:
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+
+type Props = {|
+  children?: React.Element<any>,
```
How about `React.Node`? Ah, no. Only after #3551.

willdurand commented on tests/unit/helpers.js:
```diff
@@ -383,3 +383,15 @@ export function createUserProfileResponse({
+// }
+export const getFakeConfig = (params = {}) => {
+  return Object.assign(configUtil.cloneDeep(config), params);
+};
```
coooool! 👍 👍 👍

kumar303 commented:
Thanks. I added the new styles and also tested this on device. The Flag link may need to be a bigger tap target but otherwise it seems to be working.
